File: 백준/Silver/11279. 최대 힙/최대 힙.cc
#include <stdio.h>
#include <iostream>
#include <queue>
using namespace std;

int main(void) {
	int n, in;
	priority_queue<int> q;
	cin >> n;
	while (n--) {
		scanf("%d", &in);
		if (!in) {
			if (q.empty()) printf("0\n");
			else {
				printf("%d\n", q.top());
				q.pop();
			}
		}
		else q.push(in);
	}	
	return 0;
}

File: 백준/Silver/11279. 최대 힙/최대 힙.py
from queue import PriorityQueue
import sys
print = sys.stdout.write
input = sys.stdin.readline
n = int(input())
pq = PriorityQueue()

for i in range(n):
    req = int(input())
    if req == 0:
        if pq.empty():
            print('0\n')
        else:
            temp = pq.get()
            print(str(temp[1]) + '\n')
    else:
        pq.put((-req, req))